### ISSUE TITLE
[TECHDEBT] Improve platform overrides organization

### ISSUE THREAD

**Issue by kalemas** [bibletime/bibletime#365]
# Summary
Inspecting the code quite often one may encounter with a lot of `#ifndef` directivities, that significantly decrease readability of the code. The good example of the problem is `directory.cpp`, that is hard to read and it seem that Mobile and Mini frontends are also mixed to the logic of that object.

# Consequences
There is no another approach for Mini and Mobile frontends to share the code rather than duplicate code (that is also bad approach). So this mostly affects those frontends, but organizing platform related code (Windows, Mac, Linux specifics) and **separate base logics from specific** should also improve readability.

# Remediation approach
Suggest to develop an extra level of abstraction from platform for Application class. It just may have a bunch of methods that may be used by platform dependent implementation. So if this style to resolve the problem will be selected it will be much easier to implement frontends:

```cpp
class BaseApplication : piblic QApplication {
    QDir getDir(const QString type) {
        if (type == "sword") {
            return QDir(getDir("user").currentPath().append("/.sword"));
        } else if (type == "user") {
            return QDir::home();
        }
    }
}
```
So Android implementation will look like this
```cpp
class Application : piblic BaseApplication {
    QDir getDir(const QString type) {
        if (type == "user") {
            return QDir("/sdcard");
        }
        return BaseApplication::getDir(type);
    }
}
```
As for Desktop frontend it would be considered as organization improvement that also result in better organized and readable code.

**Comment by jaakristioja** [comment]
I agree. Pull request(s), please. :)

**Comment by kalemas** [comment]
Oh, is a looong story... There should be a bunch of prototypes, i should decide how to define final implementation (cmake defines, another object with ifndefs or something else, maybe there are suitable cpp patterns), should it compose from mixins or not, there should be a strategy to integrate this change. 
So i will proceed thinking))

**Comment by jaakristioja** [comment]
Yes. Feel free to assign yourself if and once you think you want to do this. There's no hurry, so please don't feel pressured. :)

**Comment by kalemas** [comment]
I hope to fix other issues, for now i will thinking and investigating, once i start resolve this i will self assign.

**Comment by jaakristioja** [comment]
I made a bunch of changes to `directory.cpp` to improve the platform overrides situation and it looks a slightly better now.
